### PR TITLE
PascalVOC openset

### PR DESCRIPTION
**Pull request by nikita-klsh**
This PR proposes to add `PascalVOC` openset.
 By now it processes Segmentation and Classification tasks. Each task is presented by the specific class, i.e `PascalSegmentation` for Segmentation task etc. 

The total list of tasks:

1. Classification
2. Detection
3. Segmentation
4. Action Classification Task
5. Boxless Action Classification
6. Person Layout

### CODE REVIEW

File: batchflow/opensets/pascal.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+    Notes
+    -----
+    - Each track contains only the subset of the total images with labels provided.
+    """
+    SOURCE_URL = 'http://host.robots.ox.ac.uk/pascal/VOC/voc2012/VOCtrainval_11-May-2012.tar'
+    sets_path = 'VOCdevkit/VOC2012/ImageSets'
+    task = None
+
+    def __init__(self, *args, bar=False, preloaded=None, train_test=True, **kwargs):
+        self.bar = tqdm.tqdm(total=2) if  bar else None
```
Comment (dpodvyaznikov) [inline batchflow/opensets/pascal.py]: Remove extra space after `if`

File: batchflow/opensets/pascal.py [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+                    'diningtable', 'dog', 'horse', 'motorbike', 'person', 'pottedplant', 'sheep', 'sofa', 'train',
+                    'tvmonitor']
+
+    def _is_class_file(self, file):
+        """ Whether file contains labels for specific class. 20 files total. """
+        return basename(dirname(file.name)) == self.task and '_trainval.txt' in file.name
+
+    def download(self, path):
+
+        def _name(file):
```
Comment (dpodvyaznikov) [inline batchflow/opensets/pascal.py]: You use this code twice in `BasePascal`. Maybe it would be better to move it to parent class?

File: batchflow/opensets/pascal.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+
+            return (train_images, train_masks), (test_images, test_masks)
+
+
+class PascalClassification(BasePascal):
+    """ Contains 11540 images and corresponding classes
+
+    Notes
+    -----
+    - Labels are provided by the vector size 20. '1' stands for the presence atleast of one object from
```
Comment (dpodvyaznikov) [inline batchflow/opensets/pascal.py]:
```suggestion
    - Labels are represented by the one-hot vector of size 20. '1' stands for the presence of at least one object from
```

File: batchflow/opensets/pascal.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+    def download_archive(self, path=None):
+        """ Download archive if not yet on the disk """
+        if path is None:
+            path = tempfile.gettempdir()
+        filename = os.path.basename(self.SOURCE_URL)
+        localname = os.path.join(path, filename)
+
+        if not os.path.isfile(localname):
+            logger.info("Downloading %s", filename)
+            urllib.request.urlretrieve(self.SOURCE_URL, localname)
```
Comment (roman-kh) [inline batchflow/opensets/pascal.py]: Since the dataset is very large add a progress bar for archive downloading.

File: batchflow/opensets/pascal.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+    Notes
+    -----
+    - Each track contains only the subset of the total images with labels provided.
+    """
+    SOURCE_URL = 'http://host.robots.ox.ac.uk/pascal/VOC/voc2012/VOCtrainval_11-May-2012.tar'
+    sets_path = 'VOCdevkit/VOC2012/ImageSets'
+    task = None
+
+    def __init__(self, *args, bar=False, preloaded=None, train_test=True, **kwargs):
```
Comment (roman-kh) [inline batchflow/opensets/pascal.py]: Since the dataset is very large allow for storing it on a disk and working with files as well.

File: batchflow/opensets/pascal.py [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+               and basename(dirname(member.name)) == 'SegmentationClass'
+
+    def download(self, path):
+        localname = self.download_archive(path)
+        with tarfile.open(localname, "r") as archive:
+            files_in_archive = archive.getmembers()
+            train_ids = self._get_ids(archive, 'train')
+            test_ids = self._get_ids(archive, 'val')
+
+            train_images = np.array([self._extract(archive, file) for file in files_in_archive
```
[the remark this answers is not in the document]
Comment (nikita-klsh) [reply batchflow/opensets/pascal.py]: As far as i know `tafile` does not provide such functionality.   
Its possible only in case we extracting the whole archive.

[the remark this answers is not in the document]
Comment (roman-kh) [reply batchflow/opensets/pascal.py]: You know the archive structure in advance. So you can easily read the list of file names from one file (e.g. `/VOCdevkit/VOC2012/ImageSets/Segmentation/train.txt`), build the list of image files needed and then load only them (i.e. without digging through the whole archive).

File: batchflow/opensets/pascal.py [hunk 7]
```diff
@@ -32,92 +31,88 @@ class BasePascal(ImagesOpenset):
...
-
+        return list_ids[:-1]
 
 class PascalSegmentation(BasePascal):
     """ Contains 2913 images and masks.
 
     Notes
     -----
-    - Index 0 corresponds to background and index 255 corresponds to 'void' or unlabelled.
+    Index 0 corresponds to background and index 255 corresponds to 'void' or unlabelled.
```
Comment (roman-kh) [inline batchflow/opensets/pascal.py]: Label 0 or class 0

File: batchflow/opensets/pascal.py [hunk 8]
```diff
@@ -32,92 +31,88 @@ class BasePascal(ImagesOpenset):
...
     """
     task = 'Segmentation'
-    name_classes = ['background', 'aeroplane', 'bicycle', 'bird', 'boat', 'bottle', 'bus', 'car', 'cat', 'chair',
-                    'cow', 'diningtable', 'dog', 'horse', 'motorbike', 'person', 'pottedplant', 'sheep', 'sofa',
-                    'train', 'tvmonitor', 'void']
-
-    def _is_train_mask(self, file, train_ids):
-        return self._name(file) in train_ids and basename(dirname(file.name)) == 'SegmentationClass'
+    classes = ['background', 'aeroplane', 'bicycle', 'bird', 'boat', 'bottle', 'bus', 'car', 'cat', 'chair',
+               'cow', 'diningtable', 'dog', 'horse', 'motorbike', 'person', 'pottedplant', 'sheep', 'sofa',
```
Comment (roman-kh) [inline batchflow/opensets/pascal.py]: Why did you remove the spaces (e.g. in 'dining table')? Get them back. It would be convenient to use these names in visualisation

File: batchflow/opensets/pascal.py [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+            with open(localname, 'wb') as f:
+                for chunk in tqdm.tqdm(r.iter_content(chunk_size=chunk_size), total=num_bars, unit='KB',
+                                       desc=filename, leave=True):
+                    f.write(chunk)
+
+    def _name(self, path):
+        """ Return file name without format """
+        return basename(path).split('.')[0]
+
+    def _imagepath(self, name):
```
Comment (roman-kh) [inline batchflow/opensets/pascal.py]:
```suggestion
    def _image_path(self, name):
```

File: batchflow/opensets/pascal.py [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+        """ Return the path to the .jpg image in the archive by its name """
+        return os.path.join(dirname(self.SETS_PATH), 'JPEGImages', name + '.jpg')
+
+    def _extract(self, archive, file):
+        data = archive.extractfile(file).read()
+        return PIL.Image.open(BytesIO(data))
+
+    def _get_ids(self, archive, part):
```
Comment (roman-kh) [inline batchflow/opensets/pascal.py]:
```suggestion
    def _extract_ids(self, archive, part):
```

File: batchflow/opensets/pascal.py [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+
+    def _name(self, path):
+        """ Return file name without format """
+        return basename(path).split('.')[0]
+
+    def _imagepath(self, name):
+        """ Return the path to the .jpg image in the archive by its name """
+        return os.path.join(dirname(self.SETS_PATH), 'JPEGImages', name + '.jpg')
+
+    def _extract(self, archive, file):
```
Comment (roman-kh) [inline batchflow/opensets/pascal.py]:
```suggestion
    def _extract_image(self, archive, file):
```

File: batchflow/opensets/pascal.py [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+    Notes
+    -----
+    Index 0 corresponds to background and index 255 corresponds to 'void' or unlabelled.
+    """
+    task = 'Segmentation'
+    classes = ['background', 'aeroplane', 'bicycle', 'bird', 'boat', 'bottle', 'bus', 'car', 'cat', 'chair',
+               'cow', 'diningtable', 'dog', 'horse', 'motorbike', 'person', 'pottedplant', 'sheep', 'sofa',
+               'train', 'tvmonitor']
+
+    def _maskpath(self, name):
```
Comment (roman-kh) [inline batchflow/opensets/pascal.py]:
```suggestion
    def _mask_path(self, name):
```

File: batchflow/opensets/pascal.py [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+
+            return (train_images, train_masks), (test_images, test_masks)
+
+
+class PascalClassification(BasePascal):
+    """ Contains 11540 images and corresponding classes
+
+    Notes
+    -----
+    - Labels are represented by the one-hot vector of size 20. '1' stands for the presence of at least one object from
```
Comment (roman-kh) [inline batchflow/opensets/pascal.py]:
```suggestion
    Labels are represented by the one-hot vector of size 20. '1' stands for the presence of at least one object from
```

Comment (roman-kh) [reply batchflow/opensets/pascal.py]: It's not `one-hot`

File: batchflow/opensets/pascal.py [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+                    value = int(row.split()[1])
+                    d[key].append(value)
+
+            train_ids = self._get_ids(archive, 'train')
+            test_ids = self._get_ids(archive, 'val')
+
+            train_images = np.array([self._extract(archive, self._imagepath(name)) for name in train_ids], dtype=object)
+            test_images = np.array([self._extract(archive, self._imagepath(name)) for name in test_ids], dtype=object)
+
+            train_labels = np.array([d[self._name(name)] for name in train_ids])
```
Comment (roman-kh) [inline batchflow/opensets/pascal.py]: Many model would gone mad with these labels. Shouldn't we transform them into more sensible format?
